Feature singly linked list descriptions enhancements

TheeMattOliver opened the pull request:
<!-- freeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->
<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->
<!-- Make sure that your PR is not a duplicate -->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->
- [x] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX

#### Description
<!-- Describe your changes in detail -->

These changes target the Singly Linked List Data Structure challenge.

Specifically, they deal with enhancing the descriptions to improve readability and aid in comprehension, as the instructions and descriptions were unclear. Grammar errors were corrected, and word choice improved.

The "helper" methods for 'size()' and 'head()' were removed as these were redundant and unnecessary, and the tests for each section were updated to reflect this change. Wherever a test called for the size or head of a list, the length and head properties were used respectively.

raisedadead commented:
@TheeMattOliver this needs a rebase and squash, to resolve the conflicts. Can you update the pull request?

Reach us in the [Contributors Chat room](https://gitter.im/FreeCodeCamp/Contributors), if you need any assistance.

TheeMattOliver committed:
fix: edit descriptions for all singly linked list challenges

editing descriptions for all singly linked list challenge sections for clarity and removing head() and size() functions as these are redundant

removing empty translation blocks

deleting whitespace, adding <hr>'s, wrapping necessary elements in <code>...</code> and <dfn>...</dfn> wherever necessary and fixing grammar and editing for clarity

fixing comma

squashed last 4 commits

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -893,68 +893,119 @@
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Please revert this line of change. We do not need to add empty translation blocks.

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -893,68 +893,119 @@
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Please revert this line of change. We do not need to add empty translation blocks.

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -988,36 +1031,70 @@
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Please revert this line of change. We do not need to add empty translation blocks.

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -1051,49 +1120,87 @@
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Please revert this line of change. We do not need to add empty translation blocks.

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -1167,71 +1278,32 @@
-      "type": "waypoint",
-      "challengeType": 1,
-      "translations": {}
+      "translations": {
```
Please revert this line of change. We do not need to add empty translation blocks.

raisedadead reviewed (changes requested):
Please remove the additional translation blocks that are empty.

TheeMattOliver commented:
Done. Thanks.

TheeMattOliver commented:
@raisedadead Anything else I need to do?

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -893,50 +893,50 @@
       "type": "waypoint",
       "challengeType": 1,
-      "translations": {}
+      "translations": {}      
```
Remove white space, refer out contributing guide to configure the linter for your text editor.

TheeMattOliver replied:
Awesome thanks so much Mrugesh @raisedadead! I committed my changes and if there's anything else I need to do just let me know!

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -865,10 +865,10 @@
+        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line: they know who they are and they can only see the next person in line, but they are not aware ...
         "Instructions",
-        "In the code, we've create two nodes, Kitten and Puppy, and we've manually connected the Kitten node to the Puppy node.",
+        "In our code editor, we've created two nodes, Kitten and Puppy, and we've manually connected the Kitten node to the Puppy node.",
```
Wrap the words like Kitten, Puppy, Cat, Dog in `<code>...</code>` tags for legibility.

raisedadead replied:
You would want to do this at all places in this pull request, where you are referring variable, keywords, etc. in the challenge seed.

raisedadead replied:
A good place to refer the formatting is :
https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json

raisedadead replied:
You many need to add `<dfn></dfn>` to definations as well, just refer the file above

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -865,10 +865,10 @@
-        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line, they know who they are and they can only see the next person in line, but they are not aware ...
+        "Another common data structure you'll run into in computer science is the linked list. A linked list is a linear collection of data elements, called nodes, each of which points to the next. Each node in a linked list contains two key pieces of information: the element itself, and a reference to the next node.",
+        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line: they know who they are and they can only see the next person in line, but they are not aware ...
         "Instructions",
```
Does not concern your pull request, but would you mind changing this to `<hr>`

```diff
- "Instructions",
+ "<hr>",
```

raisedadead reviewed (changes requested):
@TheeMattOliver 

Thanks a lot for taking the time to contribute, we really appreciate the efforts.
I have left some comments, basically related to the formatting.
When you have done addressing the changes, please [squash your commits](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231).
Will take a look and merge, thanks once again

TheeMattOliver commented:
@raisedadead I had a syntax error. Fixed and squashed.

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -865,11 +865,12 @@
-        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line, they know who they are and they can only see the next person in line, but they are not aware ...
+        "Another common data structure you'll run into in computer science is the <dfn>linked list</dfn>. A linked list is a linear collection of data elements, called 'nodes', each of which points to the next. Each <dfn>node</dfn> in a linked list contains two key pieces of information: the <code>element</code> itself, and a reference to the next <code>node</code>.",
+        "Imagine that you are in a conga line. You have your hands on the next person in the line, and the person behind you has their hands on you. You can see the person straight ahead of you, but they are blocking the view of the other people ahead in line. A node is just like a person in a conga line: they know who they are and they can only see the next person in line, but they are not aware ...
         "Instructions",
```
Delete this line.

```diff
- "Instructions",
```

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -899,44 +900,45 @@
+        "We'll want to have a way to add items to our linked list, so the first method we'll want to create is the <code>add</code> method.", 
+        "If our list is empty, adding an element to our linked list is straightforward enough: we just wrap that element in a <code>Node</code> class, and we assign that node to the <code>head</code> of our linked list." ,
+        "But what if our list already has one or more members? How do we add an element to the list? Recall that each node in a linked list has a <code>next</code> property. To add a node to the list, find the last node in the list, and point that last node's <code>next</code> property at our new node. (Hint: you know you've reached the end of a linked list when a node's <code>next</code> propert ...
         "Instructions",
```
Delete this line.

```diff
- "Instructions",
```

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -948,13 +950,15 @@
+        "Whenever we remove a node from a linked list, it's important that we don't accidentally orphan the rest of the list in doing so. Recall that every node's <code>next</code> property points to the node that follows it in the list. If we're removing the middle element, say, we'll want to make sure that we have a connection from that element's previous node's <code>next</code> property to th ...
+        "This might sound really confusing, so let's return to the conga line example so we have a good conceptual model. Picture yourself in a conga line, and the person directly in front of you leaves the line. The person who just left the line no longer has her hands on anyone in line--and you no longer have your hands on the person that left. You step forward and put your hands on next person ...
+        "If the element we wish to remove is the <code>head</code> element, we reassign the <code>head</code> to the second node of the linked list.",
         "Instructions",
```
Delete this line.

```diff
- "Instructions",
```

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -1014,10 +1018,12 @@
+        "Let's add a few more useful methods to our linked list class. Wouldn't it be useful if we could tell if our list was empty or not, as with our <code>Stack</code> and <code>Queue</code> classes?",
+        "We should also be able to find specific elements in our linked list. Traversing through data structures is something you'll want to get a lot of practice with! Let's create an <code>indexOf</code> method that takes an <code>element</code> as an argument, and returns that element's <code>index</code> in the linked list. If the element is not found in the linked list, return <code>-1</code ...
+        "Let's also implement a method that does the opposite: an <code>elementAt</code> method that takes an <code>index</code> as an argument and returns the <code>element</code> at the given <code>index</code>. If no <code>element</code> is found, return <code>undefined</code>.",
         "Instructions",
```
Delete this line.

```diff
- "Instructions",
```

raisedadead commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -1093,11 +1101,15 @@
+        "Let's write a <code>removeAt</code> method that removes the <code>element</code> at a given <code>index</code>. The method should be called <code>removeAt(index)</code>. To remove an <code>element</code> at a certain <code>index</code>, we'll need to keep a running count of each node as we move along the linked list.",
+        "A common technique used to iterate through the elements of a linked list involves a <dfn>'runner'</dfn>, or sentinel, that 'points' at the nodes that your code is comparing. In our case, starting at the <code>head</code> of our list, we start with a <code>currentIndex</code> variable that starts at <code>0</code>. The <code>currentIndex</code> should increment by one for each node we pas ...
+        "Just like our <code>remove(element)</code> method, we need to be careful not to orphan the rest of our list when we remove the node in our removeAt(index) method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node.",
         "Instructions",
```
Delete this line.

```diff
- "Instructions",
```

raisedadead commented:
@TheeMattOliver Great work so far, almost there. Some final touches and we are good to go. Sorry for being picky.. 😅 !

when you have made the changes please amend the commit like this:

```sh
git commit --amend --no-edit
git push origin --force
```

This will let you edit the commit, without creating more.
Good luck, and thanks a lot for helping us out.

If you need any assistance hit us in the [Contributors Chat room](https://gitter.im/FreeCodeCamp/Contributors).

TheeMattOliver commented:
Done! Thanks @raisedadead. Let me know if you'd like me to make further changes.

raisedadead commented:
@TheeMattOliver looks like you forgot to do a `git add --all`, the changes are not reflecting.

```sh
git add --all
git commit --amend --no-edit
git push origin --force
```

TheeMattOliver commented:
Thanks @raisedadead I was wondering about that. I followed your last instructions literally. 

Just pushed again.

👍

raisedadead commented:
@TheeMattOliver thanks for making the changes. Will merge after a QA check first thing in the morning.

TheeMattOliver commented:
Hey guys did this get hung up on a check I didn't pass again? @raisedadead

raisedadead commented:
@TheeMattOliver looks like this got into a merge conflict. Do not worry we will resolve and merge upstream.

raisedadead commented:
@TheeMattOliver I have resolved the conflicts, but this will break the tests. Mind having a look?

raisedadead commented:
On you local setup, you need to delete this branch (local machine only, NOT on GitHub) and pull down again.

```
git checkout staging
git branch -D feature-singly-linked-list-descriptions-enhancements
git fetch origin --all
git checkout feature-singly-linked-list-descriptions-enhancements
Make Necessary changes to the tests.
git add --all
git commit --amend --no-edit
git push origin feature-singly-linked-list-descriptions-enhancements --force
```

TheeMattOliver commented:
@raisedadead Will do now, pls hold.

raisedadead commented:
Hey @TheeMattOliver ! I guess you would have to make a fresh PR. I couldn't get this working. Sincerest apologies for the inconvenience.

Let me know if you need assistance. 

Thanks a lot once again for your contribution.
